src/atlas/meshgenerator/detail/CubedSphereDualMeshGenerator.cc

```diff
@@ -0,0 +1,392 @@
+    idx_t nTriCell = 0;
+    for ( idx_t idxCsNode = 0; idxCsNode < csNodes.size(); ++idxCsNode ) {
+
+        // Exclude outer halo ring of cubed sphere mesh nodes.
+        if ( csNodesHalo( idxCsNode ) < nHalo ) {
+
+            // Tile corner elements are triangles. Rest are quads.
+            if ( cornerCell( idxCsNode ) ) {
+                ++nTriCell;
+            }
+            else {
+                ++nQuadCell;
+            }
+        }
+    }
+
+    // Add cells to mesh.
+    auto& cells = mesh.cells();
+    cells.add( new mesh::temporary::Quadrilateral(), nQuadCell );
+    cells.add( new mesh::temporary::Triangle(), nTriCell );
+    auto& nodeConnectivity = cells.node_connectivity();
+
+    // Add extra fields.
+    auto tijField = cells.add( Field( "tij", array::make_datatype<idx_t>(),
+                                      array::make_shape( cells.size(), 3 ) ) );
+    tijField.set_variables( 3 );
+
+    Field xyField = cells.add( Field( "xy", array::make_datatype<double>(),
+                                      array::make_shape( cells.size(), 2 ) ) );
+    xyField.set_variables( 2 );
+
+    Field lonLatField = cells.add( Field( "lonlat", array::make_datatype<double>(),
+                                          array::make_shape( cells.size(), 2 ) ) );
+    lonLatField.set_variables( 2 );
+
+    // Make views to dual mesh fields.
+    auto cellsGlobalIdx = array::make_view<gidx_t, 1>( cells.global_index() );
+    auto cellsRemoteIdx = array::make_indexview<idx_t, 1>( cells.remote_index() );
+    auto cellsPart      = array::make_view<int, 1>( cells.partition() );
+    auto cellsHalo      = array::make_view<int, 1>( cells.halo() );
+    auto cellsFlags     = array::make_view<int, 1>( cells.flags() );
+    auto cellsTij       = array::make_view<idx_t, 2>( tijField );
+    auto cellsXy        = array::make_view<double, 2>( xyField );
+    auto cellsLonLat    = array::make_view<double, 2>( lonLatField );
+
+    // Set separate counters for quads and triangles.
+    idx_t idxQuadCell = cells.elements( 0 ).begin();
+    idx_t idxTriCell = cells.elements( 1 ).begin();
+
+    // Loop over mesh nodes and set dual mesh cells.
+    for ( idx_t idxCsNode = 0; idxCsNode < nQuadCell + nTriCell; ++idxCsNode ) {
+
+        const idx_t tCell = csNodesTij( idxCsNode, Coordinates::T );
+        const idx_t iCell = csNodesTij( idxCsNode, Coordinates::I );
+        const idx_t jCell = csNodesTij( idxCsNode, Coordinates::J );
+
+        idx_t idxCell;
+
+        // Set corner triangle.
+        if ( cornerCell( idxCsNode ) ) {
+            idxCell = idxTriCell++;
+
+            auto triCellNodes = std::array<idx_t, 3>{};
+
+            // Bottom-left corner.
+            if ( iCell == 0 && jCell == 0 ) {
+                triCellNodes = {
+                    csCellsFunctionSpace.index( tCell,  0, -1 ),
+                    csCellsFunctionSpace.index( tCell,  0,  0 ),
+                    csCellsFunctionSpace.index( tCell, -1,  0 )};
```

Comment:
I thought we agreed on the order of nodes of a cell to be anticlockwise, starting from the lowest global index. Is there a different rule here?

Reply:
Definitely nodes should be anti-clockwise. Which node is first should be of no concern, unless you require this knowledge in the customised CubedSphere function spaces.

Reply:
You're right! They're going the wrong way around!

Reply:
Marek is super thorough! 👌

Reply:
This is why I have so much difficulty opening and closing taps!

Reply:
Wait a sec, this _is_ going anticlockwise:
```
o o    (0, -1)
  x
  
o x    (0, 0)
  o
  
x o    (-1, 0)
  o
```

Reply:
Noooooo! Arggggggj!  I have got x and y mixed up! All is ok!

Reply:
It is ok if it is tij and not tji.